### PR TITLE
Checks for muon analysis x range

### LINKED ISSUE #24052: muon analysis settings bug
In muon analysis go to the settings tab and change the start value to greater than 32 (beyond the data). Mantid will crash.

### PR DESCRIPTION
**Pull request by AnthonyLim23**
**Description of work.**
Muon analysis was crashing Mantid when the x range was invalid. This PR is to fix it. 
**Report to:** James. 

**To test:**
1. Open Muon Analysis
2. Load some data 
3. Go to the settings tab
4. Change the time axis, start value to be greater than the end time (use to crash)
5. Change the time axis end value to 50 and then the start value to 40 (use to crash)
6.  Change the time axis end value to -10 and then the start value to -20 (use to show no data)

Fixes #24052. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/Muon/MuonAnalysis.cpp [hunk 1]
```diff
@@ -1935,11 +1935,42 @@ void MuonAnalysis::plotSpectrum(const QString &wsName, bool logScale) {
 QMap<QString, QString> MuonAnalysis::getPlotStyleParams(const QString &wsName) {
   // Get parameter values from the options tab
   QMap<QString, QString> params = m_optionTab->parsePlotStyleParams();
+  auto upper = m_uiForm.timeAxisFinishAtInput->text().toDouble();
+
+  Workspace_sptr ws_ptr =
```
Comment (DanNixon) [inline qt/scientific_interfaces/Muon/MuonAnalysis.cpp]: `Workspace_const_sptr`

File: qt/scientific_interfaces/Muon/MuonAnalysis.cpp [hunk 2]
```diff
@@ -1935,11 +1935,42 @@ void MuonAnalysis::plotSpectrum(const QString &wsName, bool logScale) {
...
+      AnalysisDataService::Instance().retrieve(wsName.toStdString());
+  MatrixWorkspace_sptr matrix_workspace =
```
Comment (DanNixon) [inline qt/scientific_interfaces/Muon/MuonAnalysis.cpp]: `MatrixWorkspace_const_sptr`

File: qt/scientific_interfaces/Muon/MuonAnalysis.cpp [hunk 3]
```diff
@@ -1935,11 +1935,42 @@ void MuonAnalysis::plotSpectrum(const QString &wsName, bool logScale) {
...
+    upper = *max_element(xData.begin(), xData.end());
+    m_uiForm.timeAxisFinishAtInput->setText(QString::number(upper));
+  }
+  params["XAxisMax"] = QString::number(upper);
+  if (lower > upper) {
+    QMessageBox::warning(this, tr("Muon Analysis"),
+                         tr("Time max is less than time min.\n"
+                            "Will change time min."),
+                         QMessageBox::Ok, QMessageBox::Ok);
+    lower = upper - 1.0; // we know it will always be lower
```
Comment (DanNixon) [inline qt/scientific_interfaces/Muon/MuonAnalysis.cpp]: Is this a more useful option that setting lower to the lowest X value?

**Review by DanNixon — CHANGES_REQUESTED** [review]
Looks to work well, just a couple of comments.

File: qt/scientific_interfaces/Muon/MuonAnalysis.cpp (hunk 1 above)
Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/MuonAnalysis.cpp]: this one cannot be const as it is used in a dynamic pointer cast

Comment (DanNixon) [reply qt/scientific_interfaces/Muon/MuonAnalysis.cpp]: The cast is to `MatrixWorkspace_sptr` which could also be `MatrixWorkspace_const_sptr`.

Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/MuonAnalysis.cpp]: yes, I now have `MatrixWorkspace_const_sptr` but line 1940 was left alone.

Comment (DanNixon) [reply qt/scientific_interfaces/Muon/MuonAnalysis.cpp]: You are still doing a non-`const` cast. Change the cast to `<const MatrixWorkspace>` and you can change `ws_ptr` to `Workspace_const_sptr`.

**Review by DanNixon — CHANGES_REQUESTED** [review]

**Comment by AnthonyLim23** [comment]
I did not realise that you could do that. Should be fixed now

**Review by DanNixon — APPROVED** [review]
Looks good to me now.

**Pull request merged**